### DISCUSSION TITLE
Should the IClasspathEntry.isTest() be returning true for runtime scoped dependencies?

### DISCUSSION THREAD

**Discussion by stephendnicholas** [eclipse-m2e/m2e-core#1669] [Q&A]
Hi There,

I'm currently writing an Eclipse plugin that wants to determine the runtime classpath of some projects in the workspace, excluding any test related entries. 

However I've noticed that the `ClasspathEntry` objects coming back from the `MavenClasspathContainer` all seem to return true from `isTest()` where the scope of the dependency is runtime or test, but not compile

E.g. Based on the toString of the 'ClasspathEntry' objects when debugging (Slightly obfuscated for work reasons):
- `C:\Users\stephenn\.m2\repository\AJAR.jar[CPE_LIBRARY][K_BINARY][sourcePath:C:/Users/stephenn/.m2/repository/AJAR-sources.jar][isExported:false][attributes:maven.pomderived=true,maven.groupId=com.maven.artifactId=ajar-api,maven.version=5.0.9-alfa1001,maven.scope=compile]`
- `C:\Users\stephenn\.m2\repository\BJAR.jar[CPE_LIBRARY][K_BINARY][sourcePath:C:/Users/stephenn/.m2/repository/BJAR-sources.jar][isExported:false][attributes:maven.pomderived=true,test=true,maven.groupId=com.maven.artifactId=bjar-runtime,maven.version=1445,maven.scope=runtime]`
- `C:\Users\stephenn\.m2\repository\com\CJAR.jar[CPE_LIBRARY][K_BINARY][sourcePath:C:/Users/stephenn/.m2/repository/CJAR-sources.jar][isExported:false][attributes:maven.pomderived=true,test=true,maven.groupId=com.maven.artifactId=cjar-testsupport,maven.version=1021,maven.scope=test]`

I guess I can look to filter on both the `maven.scope` and the `test` attributes (if test=true, then also check maven.scope is not runtime), but I was wondering if it might be a bug that `maven.scope=runtime` seems to always have `test=true`? Or perhaps I'm just misunderstanding how these two are supposed to interact?

Hope that makes sense

Thanks,

Steve

**Comment by laeubi** [comment]
> I was wondering if it might be a bug that maven.scope=runtime seems to always have test=true

this really seems not to be correct, do you like to propose a PR to change that and we will see if something fails? I would even argue that the check should be the other way round, only mark it as test if scope == test and in all other cases add it to the default classpath.

**Reply by stephendnicholas** [reply]
Thanks for your comment

I've checked out out the code and had a look, and I believe:
- This is the line which is causing the `test=true`: https://github.com/eclipse-m2e/m2e-core/blame/6880fa1dae00ea437b4c13d5a39f4832af20b96d/org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/DefaultClasspathManagerDelegate.java#L182
- Which tracks back to this change: https://github.com/eclipse-m2e/m2e-core/commit/a7c65e3f08584932ba5ef2c8428239e7d0570dc9
- And this bug: https://bugs.eclipse.org/bugs/show_bug.cgi?id=414645

So it appears this was a purposeful change, and looks to have been made so that runtime scope dependencies aren't showing in the compile classpath

Having a closer look at [`IClasspathAttribute.TEST`](https://github.com/eclipse-jdt/eclipse.jdt.core/blob/master/org.eclipse.jdt.core/model/org/eclipse/jdt/core/IClasspathAttribute.java#L258-L260), the javadoc states:

```
         * During the compilation of main sources, only code is visible, that is reachable via classpath entries which do
	 * not have the test attribute set to to "true". During the compilation of test sources, all code is visible as if
	 * this attribute didn't exist at all.
```

Which makes me think this might be correct, and it is my understanding that is incorrect, around `TEST` (which underpins `isTest()`) and it's relationship to runtime scope

What do you think?

**Reply by fbricon** [reply]
Indeed this is 100% intentional. See https://bugs.eclipse.org/bugs/show_bug.cgi?id=414645#c4

**Reply by laeubi** [reply]
@fbricon thanks, maybe we should add this to the code as its not obvious but

> A quick check showed that simply adding the "test" classpath attribute to runtime dependencies:
> - prevents main code to depend on those dependencies
> - allows test code to depend on those dependencies
> - keeps those dependencies added at runtime, so that Run As still works as expected

[maps to this](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#dependency-scope):

> **runtime**
> This scope indicates that the dependency is not required for compilation, but is for execution. Maven includes a dependency with this scope in the runtime and test classpaths, but not the compile classpath.

I'm just wondering about the last part:
> - keeps those dependencies added at runtime, so that Run As still works as expected

is this a m2e feature or something generic to run configurations?
